src/write_open_telemetry.cc

```diff
@@ -0,0 +1,275 @@
+ *     Path "/v1/metrics"
+ *   </Node>
+ * </Plugin>
+ */
+
+extern "C" {
+#include "collectd.h"
+
+#include "plugin.h"
+#include "utils/common/common.h"
+
+#include "utils/resource_metrics/resource_metrics.h"
+#include "utils/strbuf/strbuf.h"
+#include "utils_complain.h"
+
+#include <netdb.h>
+}
+
+#include <grpc++/grpc++.h>
+
+#include "opentelemetry/proto/collector/metrics/v1/metrics_service.grpc.pb.h"
+#include "utils/format_open_telemetry/format_open_telemetry.h"
+
+#ifndef OT_DEFAULT_HOST
+#define OT_DEFAULT_HOST "localhost"
+#endif
+
+#ifndef OT_DEFAULT_PORT
+#define OT_DEFAULT_PORT "4317"
+#endif
+
+using opentelemetry::proto::collector::metrics::v1::
+    ExportMetricsServiceResponse;
+using opentelemetry::proto::collector::metrics::v1::MetricsService;
+
+/*
+ * Private variables
+ */
+typedef struct {
+  char *name;
+  int reference_count;
+
+  char *host;
+  char *port;
+
+  resource_metrics_set_t resource_metrics;
+  cdtime_t staged_time;
+
+  std::unique_ptr<MetricsService::Stub> stub;
+
+  pthread_mutex_t mu;
+} ot_callback_t;
+
+static int export_metrics(ot_callback_t *cb) {
+  if (cb->stub == NULL) {
+    strbuf_t buf = STRBUF_CREATE;
+    strbuf_printf(&buf, "%s:%s", cb->host, cb->port);
+
+    auto chan =
+        grpc::CreateChannel(buf.ptr, grpc::InsecureChannelCredentials());
+    cb->stub = MetricsService::NewStub(chan);
+
+    STRBUF_DESTROY(buf);
+  }
+
+  auto req = format_open_telemetry_export_metrics_service_request(
+      cb->resource_metrics);
+
+  grpc::ClientContext context;
+  ExportMetricsServiceResponse resp;
+  grpc::Status status = cb->stub->Export(&context, *req, &resp);
+
+  if (!status.ok()) {
+    ERROR("write_open_telemetry plugin: Exporting metrics failed: %s",
+          status.error_message().c_str());
+    return -1;
+  }
+
+  if (resp.has_partial_success() &&
+      resp.partial_success().rejected_data_points() > 0) {
+    auto ps = resp.partial_success();
+    NOTICE("write_open_telemetry plugin: %" PRId64
+           " data points were rejected: %s",
+           ps.rejected_data_points(), ps.error_message().c_str());
+  }
+
+  delete req;
+  return 0;
+}
+
+/* NOTE: You must hold cb->send_lock when calling this function! */
+static int ot_flush_nolock(cdtime_t timeout, ot_callback_t *cb) {
+  if (cb->resource_metrics.num == 0) {
+    cb->staged_time = cdtime();
+    return 0;
+  }
+
+  /* timeout == 0  => flush unconditionally */
+  if (timeout > 0) {
+    cdtime_t now = cdtime();
+    if (now < (cb->staged_time + timeout))
+      return 0;
+  }
+
+  int status = export_metrics(cb);
+  resource_metrics_reset(&cb->resource_metrics);
+
+  return status;
+}
+
+static void ot_callback_decref(void *data) {
+  ot_callback_t *cb = (ot_callback_t *)data;
+  if (cb == NULL)
+    return;
+
+  pthread_mutex_lock(&cb->mu);
+  cb->reference_count--;
+  if (cb->reference_count > 0) {
+    pthread_mutex_unlock(&cb->mu);
+    return;
+  }
+
+  ot_flush_nolock(/* timeout = */ 0, cb);
+
+  cb->stub.reset();
```

Comment:
MetricsService stub is just reset, it and channel gets never freed?  Would that need some suppression for Valgrind memcheck leak checker?

Reply:
If you enable `--leak-check=full`, does that "definitely lost" originate from `collectd`?

Reply:
I fixed a minor leak (e93000d724f43c6718c3665c4254dd7d39b66e0c, the callback's name wasn't freed). The remainder are outside of our control:

```
==172573== 40 bytes in 1 blocks are definitely lost in loss record 108 of 227
==172573==    at 0x48407B4: malloc (vg_replace_malloc.c:381)
==172573==    by 0x4002447: malloc (rtld-malloc.h:56)
==172573==    by 0x4002447: _dl_map_object_deps (dl-deps.c:422)
==172573==    by 0x400B8EC: dl_open_worker_begin (dl-open.c:592)
==172573==    by 0x49DEE09: _dl_catch_exception (dl-error-skeleton.c:208)
==172573==    by 0x400B045: dl_open_worker (dl-open.c:782)
==172573==    by 0x49DEE09: _dl_catch_exception (dl-error-skeleton.c:208)
==172573==    by 0x400B437: _dl_open (dl-open.c:884)
==172573==    by 0x4915347: dlopen_doit (dlopen.c:56)
==172573==    by 0x49DEE09: _dl_catch_exception (dl-error-skeleton.c:208)
==172573==    by 0x49DEEBE: _dl_catch_error (dl-error-skeleton.c:227)
==172573==    by 0x4914E36: _dlerror_run (dlerror.c:138)
==172573==    by 0x49153F8: dlopen_implementation (dlopen.c:71)
==172573==    by 0x49153F8: dlopen@@GLIBC_2.34 (dlopen.c:81)
==172573==
==172573== 2,816 bytes in 8 blocks are possibly lost in loss record 212 of 227
==172573==    at 0x484220F: operator new[](unsigned long) (vg_replace_malloc.c:640)
==172573==    by 0x63BE3FF: ??? (in /usr/lib/x86_64-linux-gnu/libabsl_random_internal_pool_urbg.so.20220623.0.0)
==172573==    by 0x5CEE4CE: void absl::debian3::base_internal::CallOnceImpl<void (&)()>(std::atomic<unsigned int>*, absl::debian3::base_internal::SchedulingMode, void (&)()) (in /usr/lib/x86_64-linux-gnu/libabsl_malloc_internal.so.20220623.0.0)
==172573==    by 0x63BE5F7: ??? (in /usr/lib/x86_64-linux-gnu/libabsl_random_internal_pool_urbg.so.20220623.0.0)
==172573==    by 0x63BE7E0: absl::debian3::random_internal::RandenPool<unsigned char>::Fill(absl::debian3::Span<unsigned char>) (in /usr/lib/x86_64-linux-gnu/libabsl_random_internal_pool_urbg.so.20220623.0.0)
==172573==    by 0x5902613: grpc_core::BackOff::BackOff(grpc_core::BackOff::Options const&) (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172573==    by 0x57E2000: grpc_core::PollingResolver::PollingResolver(grpc_core::ResolverArgs, grpc_core::ChannelArgs const&, grpc_core::Duration, grpc_core::BackOff::Options, grpc_core::TraceFlag*) (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172573==    by 0x57C560C: ??? (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172573==    by 0x57C5CD1: ??? (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172573==    by 0x59A63DA: grpc_core::ResolverRegistry::CreateResolver(absl::debian3::string_view, grpc_core::ChannelArgs const&, grpc_pollset_set*, std::shared_ptr<grpc_core::WorkSerializer>, std::unique_ptr<grpc_core::Resolver::ResultHandler, std::default_delete<grpc_core::Resolver::ResultHandler> >) const (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172573==    by 0x572099E: grpc_core::ClientChannel::CreateResolverLocked() (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172573==    by 0x5720CF1: grpc_core::ClientChannel::TryToConnectLocked() (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
```

Reply:
You could try e.g. `--num-callers=30` to see whether "possibly lost" in _grpc_ leads to something that could be  due to something `collectd` does.

Reply:
Here you go:

```
==172707== 2,816 bytes in 8 blocks are possibly lost in loss record 212 of 227
==172707==    at 0x484220F: operator new[](unsigned long) (vg_replace_malloc.c:640)
==172707==    by 0x63BE3FF: ??? (in /usr/lib/x86_64-linux-gnu/libabsl_random_internal_pool_urbg.so.20220623.0.0)
==172707==    by 0x5CEE4CE: void absl::debian3::base_internal::CallOnceImpl<void (&)()>(std::atomic<unsigned int>*, absl::debian3::base_internal::SchedulingMode, void (&)()) (in /usr/lib/x86_64-linux-gnu/libabsl_malloc_internal.so.20220623.0.0)
==172707==    by 0x63BE5F7: ??? (in /usr/lib/x86_64-linux-gnu/libabsl_random_internal_pool_urbg.so.20220623.0.0)
==172707==    by 0x63BE7E0: absl::debian3::random_internal::RandenPool<unsigned char>::Fill(absl::debian3::Span<unsigned char>) (in /usr/lib/x86_64-linux-gnu/libabsl_random_internal_pool_urbg.so.20220623.0.0)
==172707==    by 0x5902613: grpc_core::BackOff::BackOff(grpc_core::BackOff::Options const&) (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x57E2000: grpc_core::PollingResolver::PollingResolver(grpc_core::ResolverArgs, grpc_core::ChannelArgs const&, grpc_core::Duration, grpc_core::BackOff::Options, grpc_core::TraceFlag*) (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x57C560C: ??? (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x57C5CD1: ??? (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x59A63DA: grpc_core::ResolverRegistry::CreateResolver(absl::debian3::string_view, grpc_core::ChannelArgs const&, grpc_pollset_set*, std::shared_ptr<grpc_core::WorkSerializer>, std::unique_ptr<grpc_core::Resolver::ResultHandler, std::default_delete<grpc_core::Resolver::ResultHandler> >) const (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x572099E: grpc_core::ClientChannel::CreateResolverLocked() (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x5720CF1: grpc_core::ClientChannel::TryToConnectLocked() (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x5966FCF: grpc_core::WorkSerializer::WorkSerializerImpl::DrainQueueOwned() (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x596755B: grpc_core::WorkSerializer::Run(std::function<void ()>, grpc_core::DebugLocation const&) (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x5716A0F: ??? (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x4F6AABC: closure_impl::closure_wrapper(void*, absl::debian3::Status) (in /usr/lib/x86_64-linux-gnu/libgrpc++.so.1.51.1)
==172707==    by 0x597C685: grpc_core::ExecCtx::Flush() (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x4F55610: grpc_core::ExecCtx::~ExecCtx() (in /usr/lib/x86_64-linux-gnu/libgrpc++.so.1.51.1)
==172707==    by 0x5A427F2: grpc_call_start_batch (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x4EABC46: grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal::CallOpSendMessage, grpc::internal::CallOpRecvInitialMetadata, grpc::internal::CallOpRecvMessage<google::protobuf::MessageLite>, grpc::internal::CallOpClientSendClose, grpc::internal::CallOpClientRecvStatus>::ContinueFillOpsAfterInterception() (call_op_set.h:971)
==172707==    by 0x4EB0646: PerformOps (call.h:67)
==172707==    by 0x4EB0646: grpc::internal::BlockingUnaryCallImpl<google::protobuf::MessageLite, google::protobuf::MessageLite>::BlockingUnaryCallImpl(grpc::ChannelInterface*, grpc::internal::RpcMethod const&, grpc::ClientContext*, google::protobuf::MessageLite const&, google::protobuf::MessageLite*) (client_unary_call.h:80)
==172707==    by 0x4EA55AF: BlockingUnaryCall<opentelemetry::proto::collector::metrics::v1::ExportMetricsServiceRequest, opentelemetry::proto::collector::metrics::v1::ExportMetricsServiceResponse, google::protobuf::MessageLite, google::protobuf::MessageLite> (client_unary_call.h:50)
==172707==    by 0x4EA55AF: opentelemetry::proto::collector::metrics::v1::MetricsService::Stub::Export(grpc::ClientContext*, opentelemetry::proto::collector::metrics::v1::ExportMetricsServiceRequest const&, opentelemetry::proto::collector::metrics::v1::ExportMetricsServiceResponse*) (metrics_service.grpc.pb.cc:43)
==172707==    by 0x4EA3646: ??? (write_open_telemetry.cc:103)
==172707==    by 0x4EA423E: ot_flush_nolock (write_open_telemetry.cc:137)
==172707==    by 0x4EA423E: ot_flush_nolock (write_open_telemetry.cc:124)
==172707==    by 0x4EA423E: ??? (write_open_telemetry.cc:178)
==172707==    by 0x119CB2: plugin_flush (plugin.c:1877)
==172707==    by 0x118702: plugin_flush_timeout_callback (plugin.c:1264)
==172707==    by 0x116AD0: plugin_read_thread (plugin.c:529)
==172707==    by 0x4919043: start_thread (pthread_create.c:442)
==172707==    by 0x499887F: clone (clone.S:100)
```

`absl::debian3::base_internal::CallOnceImpl` sounds to me like this is initializing some memory at start-up (a random pool?) and never cleans it up. Letting the binary run for longer yields the same 2816 bytes.

Reply:
I agree, this can be ignored.

(It's good to know that it was `MetricsService::Stub::Export()` call in collectd that triggered though.)

Reply:
Yeah, probably lazy initialization …